### ISSUE TITLE
Make it possible to skip only one hop of __torch_function__ override

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#55093]
@rgommers @hameerabbasi and I have argued about whether or not DisableTorchFunction should be public API or not. Despite being undocumented and unsupported, there have been a few uses of it in the wild https://github.com/search?l=Python&q=disabletorchfunction&type=Code mostly for when people want to delegate to an underlying implementation.

I want to float an alternative idea for how to expose DisableTorchFunction in a user friendly way, and it stems from thinking about the following situation. Suppose you have a Python program that looks like this:

```
def foo(self):
    if has_torch_function_unary(self):
        return handle_torch_function(foo, (self,), self)
    return bar(self) + bar(self)

def bar(self):
    if has_torch_function_unary(self):
        return handle_torch_function(bar, (self,), self)
    return torch.built_in(self) * 2
```

Today, if I define a torch function, calling foo always shunts me off to the torch function. I cannot choose to *partially delegate* the implementation of foo in terms of my implementation of bar; if I re-call the passed in foo on torch function, I'll infinite loop; if I use DisableTorchFunction, I completely disable torch function handling (and just rely on the intrinsic meaning of tensor computation).  What I'd like to do is disable torch function handling *only* for foo.

This isn't really possible to do in the way we write has torch function today. But a small modification would work:

```
    if has_torch_function_unary(foo, self):
...
    if has_torch_function_unary(bar, self):
```

This would allow the has_torch_function to then selectively return false if a torch function advertised that it didn't have support for foo. Although we could do this by adding `__has_torch_function__` magic method, it might just be simpler and more obvious if we let `DisableTorchFunction(op)` take an operator (and then have `has_torch_function` query it).  The set shouldn't be too large so it should be a few pointer comparisons in the common case.

The main deal is if we do this, we'll have to update all the `has_torch_function` call sites again. We've done it before, we can probably do it again.

**Comment by rgommers** [comment]
@peterbell10 maybe this is a good issue to start with `__torch_function__`. I think @hameerabbasi is already looking at this, but collaborating could be useful as this is a large-ish design question.

**Comment by peterbell10** [comment]
If I understand, the proposed usage would be:
```
with torch.overrides.DisableTorchFunction(func):
    ret = func(*args, **kwargs)
```

The numpy spelling of that would be:
```
ret = func._implementation(*args, **kwargs)
```
where `func._implementation` doesn't disable dispatch, instead it just doesn't call any dispatch logic at all. Since dispatch isn't disabled, any functions called inside `_implementation` is still free to dispatch.

Would it be reasonable to provide a split API like this in PyTorch?

**Comment by ezyang** [comment]
I'm OK with `_implementation` too but I don't know how to actually implement this. You need a way to actually bypass the dispatch logic, but the implication of this spelling is that it is only disabled once (and if you recursively reenter the function, it would no longer be disabled). I don't see how it's implementable in this variant:

```
def foo(self):
    if has_torch_function_unary(self):
        return handle_torch_function(foo, (self,), self)
    return bar(self) + bar(self)
```

or really in any variant that doesn't split the function into two pieces (but we can't do that because we know that `has_torch_function` test is perf sensitive, see https://github.com/pytorch/pytorch/pull/48965

**Comment by hameerabbasi** [comment]
## What we could do
We could just develop another context manager like `torch._C._DisableTorchFunction` called `torch._C._DisableTorchFunctionOnce` which does the following:

1. Sets a static thread-local boolean to `false`.
2. We set it to `true` when we call `__torch_function__` ever.

## NumPy's Implementation
The implementation in NumPy is deceptively simple, but because `__torch_function__` has a different implementation to `__array_function__` for performance reasons, it's hard for `__torch_function__`. It's divided into a clean dispatch and implementation layer.

Imagine a NumPy function with array arguments and non-array arguments:

```python
def numpy_function(arr_arg, non_arr_arg, arr_kwarg=None):
    # Implementation
```

What we do is have a simple decorator:

```python
def array_function_dispatch(dispatcher):
    def wrapper(func):
        @functools.wraps(func)
        def dispatched(*a, **kw):
            relevant_args = dispatcher(*a, **kw)
            if has_array_function(*relevant_args):  # Written in C
                # Extract and order types here
                return handle_array_function(func, types, a, kw)  # Also written in C
            return func(*a, **kw)
        dispatched._implementation = func

# Used like this
def _numpy_function_dispatcher(arr_arg, non_arr_arg, arr_kwarg=None):  # Signature must match
    return arr_arg, arr_kwarg

@array_function_dispatch(_numpy_function_dispatcher)
def numpy_function(arr_arg, non_arr_arg, arr_kwarg=None):
    # Implementation
```

This way, you get dispatch for the cost of two extra Python function calls and an if (which has been discussed and isn't an option in PyTorch). The reason it's hard to write `array_function_dispatch` in C/C++ is that it's:

1. Hard to put an `_implementation` slot on it so you can stash the skipped impl somewhere.
2. The dispatcher itself is in Python, writing it in C++ might be prohibitive for Python devs.

**Comment by ezyang** [comment]
No, I still don't see how this would work.

> Sets a static thread-local boolean to false.

Let's consider what happens in the above example

```
def foo(self):
    if has_torch_function_unary(self):
        return handle_torch_function(foo, (self,), self)
    return bar(self) + bar(self)

def bar(self):
    if has_torch_function_unary(self):
        return handle_torch_function(bar, (self,), self)
    return torch.built_in(self) * 2
```

1. I assume the TLS starts as true
2. We call foo, has torch function returns true, we hit handle torch function
3. We set it again to true (by rule 2), go to torch function handler
4. Torch function handler sets TLS to false, then reinvokes function
5. We call foo again, has torch function returns false, we go to the body
6. We now call bar. What exactly do you propose should happen when we hit bar's has torch function?

The point of this example is the calling convention for Python wrapped dispatchers are different from C++; there is no guaranteed preamble/postamble which we can hit for you to do the "reset TLS to true" logic.

**Comment by hameerabbasi** [comment]
> The point of this example is the calling convention for Python wrapped dispatchers are different from C++; there is no guaranteed preamble/postamble which we can hit for you to do the "reset TLS to true" logic.


Assuming that Python will always have a refcounting GC, we can use `__init__`/`__del__` to detect deleted objects and also use the C++ RAII pattern.

MVP:
```python
>>> class X:
...     def __del__(self):
...         print("I was deleted")
...
>>> def f():
...     x = X()
...     print("I came first!")
...
>>> f()
I came first!
I was deleted
```

**Comment by peterbell10** [comment]
I think `DisableTorchFunctionOnce` should work but it's not `__torch_function__` that resets the flag, it's `has_torch_function`. If the `skip_next_torch_function` flag is true, `has_torch_function` returns `false` and resets the flag to false. 

Considering Ed's example:
1. We call `foo` with `skip_next_torch_function` initially `true`
2. `has_torch_function` resets `skip_next_torch_function` to `false` and returns `false`.
3. We continue the body of `foo`, which in turn calls `bar`
4. `has_torch_function` returns true inside of `bar` and dispatches.

The problem with this in general is it breaks if `foo` doesn't call `has_torch_function`. It should always work for redispatching inside of `__torch_function__` though.

One idea might be to not expose the context manager, but instead add an `_implementation` method that works like this:
```
def _implmentation(self, *args, **kwargs):
    with torch._C.DisableTorchFunctionOnce():
        self(*args, **kwargs)
```

**Comment by hameerabbasi** [comment]
I will add three things to that:

1. We need a flag set by the context manager to decide when to do that.
2. We can set the flag to `false` again after calling `__torch_function__` in `handle_torch_function`.
3. We can use integers instead of booleans to represent levels, in this way, the context manager can be made re-entrant.

**Comment by peterbell10** [comment]
> We can set the flag to false again after calling `__torch_function__` in `handle_torch_function`.

Doesn't `has_torch_function` always get called before `handle_torch_function`? That should mean the flag is always false if `handle_torch_function` gets called.

> We can use integers instead of booleans to represent levels, in this way, the context manager can be made re-entrant.

I don't think this works as you'd want it to. For Ed's example `bar` is called twice, so with a depth of two you'd get the second `bar` being dispatched.

**Comment by ezyang** [comment]
> I think DisableTorchFunctionOnce should work but it's not __torch_function__ that resets the flag, it's has_torch_function. If the skip_next_torch_function flag is true, has_torch_function returns false and resets the flag to false.

This is like old school, C errno style stuff, so I don't like it very much. If the setting "leaks" (you miss `has_torch_function` call, which by the way, one might naively assume is side effect free based on the name) you'll probably end up with a hard to debug problem somewhere else far away.

**Comment by hameerabbasi** [comment]
Here's a version that I've been toying in my head that is side effect free:

1. There are two integers, `torch_function_levels` and `torch_function_skip` (which I'll abbreviate), both starting out at `0`.
2. `has_torch_function` queries and returns true if both are equal.
3. `handle_torch_function` increments before `__torch_function__` and decrements after, but iff `levels > start` at the beginning of the call.
4. `DisableTorchFunctionOnce` just increments `skip`.

**Comment by hameerabbasi** [comment]
> > I think DisableTorchFunctionOnce should work but it's not __torch_function__ that resets the flag, it's has_torch_function. If the skip_next_torch_function flag is true, has_torch_function returns false and resets the flag to false.
> 
> 
> 
> This is like old school, C errno style stuff, so I don't like it very much. If the setting "leaks" (you miss `has_torch_function` call, which by the way, one might naively assume is side effect free based on the name) you'll probably end up with a hard to debug problem somewhere else far away.

This is true today, if you miss it and add the `handle_torch_function`, you get infinite recursion or equivalent in the tests. If you don't add either, the function doesn't get dispatched either way. So the net effect will be the same, unless one uses `has_torch_function` for a reason other than to redispatch, and I admit I can see that happening.

Ed's original design is okay IMO, if backcompat isn't an issue.

**Comment by peterbell10** [comment]
> If the setting "leaks" (you miss has_torch_function call, which by the way, one might naively assume is side effect free based on the name) you'll probably end up with a hard to debug problem somewhere else far away.

Yes, it's a bit hacky. Which is why I suggest not making it a public context manager public and instead providing an `_implementation` method that uses it internally. Inside of `_implementation` you can have an `assert not skip_next_torch_function`,  or a `TORCH_INTERNAL_ASSERT`.

The main issue for me is that you really want to make a single call without dispatch, not disable anything globally. If you'll allow a slightly contrived example, imagine if the custom implementation of `bar` internally calls `foo` with a different kind of custom tensor. 
```python
class CustomTensorA:
    inner_custom_tensor: CustomTensorB
    def bar(self):
        return torch.foo(inner_custom_tensor) / 2
```

Since `foo` dispatch is disabled globally, it won't be dispatched for `CustomTensorB`.

**Comment by hameerabbasi** [comment]
> One idea might be to not expose the context manager, but instead add an `_implementation` method that works like this:

This is possible, although I'm not entirely sure it can be zero performance cost. Consider:

```python
>>> torch.add._implementation = None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'builtin_function_or_method' object has no attribute '_implementation'
>>> type(torch.add)
<class 'builtin_function_or_method'>
```

This implies we have to create a custom class with a `tp_call` or something overridden, which implies massive changes to codegen. Also this new class might not be as cheap to call as a `builtin_function_or_method`.

For Python functions, this works:

```python
>>> def f():
...     pass
... 
>>> type(f) is type(torch.add)
False
>>> type(f)
<class 'function'>
>>> f._implementation = None
```

**Comment by peterbell10** [comment]
My main point is that we really want the disabling to be local to a particular call. The syntax for how we achieve that could be different from NumPy. Another possibility could be an implicit `_disable_torch_function=False` keyword-only argument.

>This implies we have to create a custom class with a `tp_call` or something overridden, which implies massive changes to codegen. Also this new class might not be as cheap to call as a `builtin_function_or_method`.

As of Python 3.8 you can (at least in theory) write custom classes that perform just as well as `builtin_function_or_method`. This is because of [PEP 590].
>This PEP proposes that the calling convention used internally for Python and builtin functions is generalized and published so that all calls can benefit from better performance.

 We implement the `vectorcall` protocol and set the `Py_TPFLAGS_METHOD_DESCRIPTOR` flag on the type. This still requires non-trivial changes to codegen but in theory, `vectorcall` has _better_ performance than `METH_VARARGS | METH_KEYWORDS` used currently. So, some changes might be justified either way.

[PEP 590]: https://www.python.org/dev/peps/pep-0590/

**Comment by ezyang** [comment]
FWIW I'm not too concerned about performance when delegating. I am now... semi convinced that @peterbell10's scheme will be OK? At least, I kind of can believe we can add enough asserts to catch us if we get it wrong.

**Comment by hameerabbasi** [comment]
> FWIW I'm not too concerned about performance when delegating. I am now... semi convinced that @peterbell10's scheme will be OK? At least, I kind of can believe we can add enough asserts to catch us if we get it wrong.

We have both ways, I'm okay with either. One thing that can be done is to have a `has_torch_function_dispatch` for dispatching and a regular `has_torch_function` for querying, without side-effects.

**Comment by peterbell10** [comment]
> `has_torch_function` call, which by the way, one might naively assume is side effect free based on the name

I found an example where `has_torch_function` is being used outside of the standard pattern, so side-affects would be unexpected:
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/distributions/utils.py#L27

Which obviously breaks if `has_torch_function` has side-effects. That said, 
1. This isn't a problem if users can only see `_implementation` methods (since `broadcast_all` wouldn't have it).
2. IMO this should really be using `torch.overrides.is_tensor_like` instead of `has_torch_function`.

If there's still concern though, we could consider naming it something other than `has_torch_function`. Something that doesn't imply it's safe to call everywhere.

Also, as a side note, are `has_torch_function` and `handle_torch_function` really meant to be part of the public API? It would be easy to extend `wrap_torch_function` to support this change, but if downstream libraries are calling `has_torch_function` and `handle_torch_function` directly then they wouldn't be able to take advantage of this without changing their own code.

**Comment by ezyang** [comment]
I'm... doubtful that this function is doing the check in the right way. It was just added to patch up a crash https://github.com/pytorch/pytorch/commit/3a0d4240c32b568ebe74a0f6901fc744d5179b3b#diff-b2c6d0aca65ff864934bbabc5a44108c927339d7f77a8093e057144ea432589f There's probably a better way to do it.

**Comment by ezyang** [comment]
Let's revive this, it is giving trouble for https://github.com/pytorch/pytorch/pull/73942

**Comment by ezyang** [comment]
@anijain2305 reports that one consequence of this is you can't actually use TorchFunctionMode on backward(), because we have backward() overridable with `__torch_function__` which forces the mode off before you actually call the backwards engine.
